src/bootstrap/src/core/build_steps/test.rs

```diff
@@ -1630,6 +1630,13 @@ struct Compiletest {
     compare_mode: Option<&'static str>,
 }
 
+/// `compiletest` under `--stage 0` causes `compiletest` to be used with the beta compiler which can
+/// cause many tests to fail. This env var is needed to indicate that the user understands the risk
+/// and actually intends to use `compiletest` with `--stage 0`.
+fn compiletest_force_stage0() -> bool {
+    env::var("COMPILETEST_FORCE_STAGE0").is_ok()
```

Comment:
Note that this just checks the presence of `COMPILETEST_FORCE_STAGE0`, it doesn't check what value it gets.

Reply:
I think checking if we are testing stage 0 is a better idea. The env var is what allows testing stage 0, not something that should affect the actual testing.

Reply:
Fixed so that bootstrap cargo is only passed when `builder.top_stage == 0`.